### PR TITLE
React: Refactor RSC out of Next

### PR DESCRIPTION
**Pull request by shilman**
Closes N/A

## What I did

The RSC approach introduced in 7.6 can apply to any React project that uses an appropriate 18.3 canary.

So this breaking change changes the feature flag & configuration parameter so that it's no longer NextJS specific.

- Feature flag: `experimentalNextRSC` => `experimentalRSC`
- Configuration parameter: `nextjs.rsc` => `react.rsc`

There is an appropriate note in `MIGRATION.md`. And since we don't actually support any other usage yet, we have not documented its use with any other framework yet.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-25591-sha-337ab526`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25591-sha-337ab526). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-25591-sha-337ab526`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25591-sha-337ab526) |
| **Triggered by** | @shilman |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`shilman/refactor-client-rsc`](https://github.com/storybookjs/storybook/tree/shilman/refactor-client-rsc) |
| **Commit** | [`337ab526`](https://github.com/storybookjs/storybook/commit/337ab52695405b06442e8f825206221d693f4812) |
| **Datetime** | Sat Jan 13 15:06:12 UTC 2024 (`1705158372`) |
| **Workflow run** | [7513198979](https://github.com/storybookjs/storybook/actions/runs/7513198979) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=25591`_
</details>

### CODE REVIEW

File: MIGRATION.md [hunk 1]
```diff
@@ -102,7 +103,7 @@
     - [Deploying build artifacts](#deploying-build-artifacts)
       - [Dropped support for file URLs](#dropped-support-for-file-urls)
       - [Serving with nginx](#serving-with-nginx)
-      - [Ignore story files from node_modules](#ignore-story-files-from-node_modules)
```
Comment (valentinpalkovic) [inline MIGRATION.md:105]: Aha! So it's actually your editor who plays a back and forth with mine :D Some formatting rules are applied here, which doesn't seem to be necessary.

File: MIGRATION.md [hunk 2]
```diff
@@ -728,6 +729,12 @@ Similar to how Next.js detects if SWC should be used, Storybook will follow more
 - If you use Next.js 14 or higher and you don't have a .babelrc file, Storybook will use SWC to transpile your code.
 - Even if you have a .babelrc file, Storybook will still use SWC to transpile your code if you set the experimental `experimental.forceSwcTransforms` flag to `true` in your `next.config.js`.
 
+##### RSC config moved to react renderer
```
Comment (valentinpalkovic) [inline MIGRATION.md:732]:
```suggestion
##### RSC config moved to React renderer
```

File: MIGRATION.md [hunk 3]
```diff
@@ -728,6 +729,12 @@ Similar to how Next.js detects if SWC should be used, Storybook will follow more
...
+
+Storybook 7.6 introduced a new feature flag, `experimentalNextRSC`, to enable React Server Components in a NextJS project. It also introduced a parameter `nextjs.rsc` to selectively disable it on particilar components or stories.
```
Comment (valentinpalkovic) [inline MIGRATION.md:734]:
```suggestion
Storybook 7.6 introduced a new feature flag, `experimentalNextRSC`, to enable React Server Components in a Next.js project. It also introduced a parameter `nextjs.rsc` to selectively disable it on particilar components or stories.
```

File: MIGRATION.md [hunk 4]
```diff
@@ -728,6 +729,12 @@ Similar to how Next.js detects if SWC should be used, Storybook will follow more
...
+
+These flags have been renamed to `experimentalRSC` and `react.rsc`, respectively. The this breaking change is to accommodate RSC support in other, non-NextJS frameworks. For now, `@storybook/nextjs` is the only framework that supports it, and does so experimentally.
```
Comment (valentinpalkovic) [inline MIGRATION.md:736]:
```suggestion
These flags have been renamed to `experimentalRSC` and `react.rsc`, respectively. This is a breaking change to accommodate RSC support in other non-Next.js frameworks. For now, `@storybook/nextjs` is the only framework that supports it, and does so experimentally.
```

File: code/renderers/react/src/entry-preview-rsc.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import type { Addon_DecoratorFunction } from '@storybook/types';
+import type { StoryContext } from './types';
+
+export const ServerComponentDecorator = (
+  Story: React.FC,
+  { parameters }: StoryContext
+): React.ReactNode => {
+  if (!parameters?.react?.rsc) return <Story />;
+
+  if (semver.major(React.version) < 18 || semver.minor(React.version) < 3) {
```
Comment (valentinpalkovic) [inline code/renderers/react/src/entry-preview-rsc.tsx:12]: This condition returns true for a hypothetical upcoming 19.0 version and will throw an error.

File: code/renderers/react/src/entry-preview-rsc.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    throw new Error('React Server Components require React 18.3');
```
Comment (valentinpalkovic) [inline code/renderers/react/src/entry-preview-rsc.tsx:13]:
```suggestion
    throw new Error('React Server Components require React >= 18.3');
```

File: MIGRATION.md [hunk 7]
```diff
@@ -728,6 +729,12 @@ Similar to how Next.js detects if SWC should be used, Storybook will follow more
 - If you use Next.js 14 or higher and you don't have a .babelrc file, Storybook will use SWC to transpile your code.
 - Even if you have a .babelrc file, Storybook will still use SWC to transpile your code if you set the experimental `experimental.forceSwcTransforms` flag to `true` in your `next.config.js`.
 
+##### RSC config moved to React renderer
+
+Storybook 7.6 introduced a new feature flag, `experimentalNextRSC`, to enable React Server Components in a Next.jsS project. It also introduced a parameter `nextjs.rsc` to selectively disable it on particular components or stories.
```
Comment (valentinpalkovic) [inline MIGRATION.md:734]:
```suggestion
Storybook 7.6 introduced a new feature flag, `experimentalNextRSC`, to enable React Server Components in a Next.js project. It also introduced a parameter `nextjs.rsc` to selectively disable it on particular components or stories.
```

File: code/frameworks/nextjs/src/preset.ts [hunk 8]
```diff
@@ -170,7 +167,7 @@ export const webpackFinal: StorybookConfig['webpackFinal'] = async (baseConfig,
     configureFastRefresh(baseConfig);
   }
 
-  if (options.features?.experimentalNextRSC) {
+  if (options.features?.experimentalRSC) {
```
Comment (tmeasday) [inline code/frameworks/nextjs/src/preset.ts:170]: Should this still be in this file?

Comment (shilman) [reply code/frameworks/nextjs/src/preset.ts:170]: It controls a webpack configuration. Where would you imagine it living?

**Pull request merged**